### DISCUSSION TITLE
Multi-select / Sortable Component - suitable libraries and final functionality

### DISCUSSION THREAD

**Discussion by AdrianFiroiu** [sixach/wp-block-components#10] [Ideas]
Hello everyone,

I have opened this discussion as part of the _wp-block-components_ repository in order to bring together a few topics from across a few PRs that touched upon this discussion's subject matter.

I would see 2 main points of discussion, with common characteristics between them, that we could address:

### 1. Multi-select Component
For the multi-select component I have noticed that we are currently using the [react-multi-select-component](https://github.com/harshzalavadiya/react-multi-select-component) package. @mahdiyazdani and @gooklani have suggested, [here](https://github.com/sixach/wp-block-components/pull/3#discussion_r639556313) and [here](https://github.com/sixach/wp-block-utils/pull/22#pullrequestreview-671816755), a more suitable alternative by using [react-select](https://github.com/jedwatson/react-select) instead.

At ~22k GitHub stars for _react-select_, compared to ~150 for _react-multi-select-component_, and backed by Atlassian it seems like the most suitable choice. One lacking feature out of the box, that I observed, was the lack of a - Select All - type of functionality, but that is something we can build ourselves quite easily.

One important aspect we've shortly discussed in regards to the multi-select component is whether or not the items within should be sortable. As @mahdiyazdani has mentioned in this [PR here](https://github.com/sixach/wp-block-components/pull/3#discussion_r639574655) - is that we should not rely on the "react-sortable-hoc" package that _react-select_ uses for demonstration but instead integrate our component with another, better supported, library such as [dnd-kit](https://github.com/clauderic/dnd-kit) or other alternatives.

❓In this direction I would like to open a few questions for clarification:
- Is this sorting within the multi-select functionality something we would like to support? (for example: does it make sense from a UX standpoint)
  - If so, should this be part of the first version of this component?
  - If so, which library should we opt for, based on point 2 below?

### 2. Sortable Component in the Inspector & library choice for sorting
For the sorting functionality, as mentioned in point 1 as well, we should not rely anymore on _react-sortable-hoc_ and choose from a few suitable libraries.

I have tested the [dnd-kit](https://github.com/clauderic/dnd-kit) package but I have found it to be a bit difficult to work with and the documentation was not the most helpful. If someone else has more experience with implementing it I would be happy to work together on it's implementation.

Because of these hurdles, I have opted for the [react-beautiful-dnd](https://github.com/atlassian/react-beautiful-dnd) package which is also backed by Atlassian with ~23k stars on GitHub. As a quick PoC, I have implemented the sorting as an `Inspector` component element below (normally we would want to split things out into separate components but in this example everything is just done in one place for simplicity):

```javascript
				
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd';

const reorder = (list, startIndex, endIndex) => {
	const result = Array.from(list);
	const [removed] = result.splice(startIndex, 1);
	result.splice(endIndex, 0, removed);

	return result;
};

const grid = 8;

const getItemStyle = (isDragging, draggableStyle) => ({
	// some basic styles to make the items look a bit nicer
	userSelect: 'none',
	padding: grid * 2,
	margin: `0 0 ${grid}px 0`,

	// change background colour if dragging
	background: isDragging ? 'lightgreen' : 'cornsilk',

	// styles we need to apply on draggables
	...draggableStyle,
});

const getListStyle = (isDraggingOver) => ({
	background: isDraggingOver ? 'lightblue' : 'grey',
	padding: grid,
	width: 250,
});


	const onDragEnd = (result) => {
		// return if item was dropped outside
		if (!result.destination) return;

		// return if the item was dropped in the same place
		if (
			result.source.droppableId === result.destination.droppableId &&
			result.source.index === result.destination.index
		  ) return;

		const items = reorder(
			postIds,
			result.source.index,
			result.destination.index,
		);

		setAttributes( {
			postIds: items,
		} );
	}

<DragDropContext onDragEnd={onDragEnd}>
	<Droppable droppableId="droppable">
	{(droppableProvided, droppableSnapshot) => (
		<div
		ref={droppableProvided.innerRef}
		style={getListStyle(droppableSnapshot.isDraggingOver)}
		>
		{postIds.map((item, index) => (
			<Draggable key={item} draggableId={item.toString()} index={index}>
			{(draggableProvided, draggableSnapshot) => (
				<div
				ref={draggableProvided.innerRef}
				{...draggableProvided.draggableProps}
				{...draggableProvided.dragHandleProps}
				style={getItemStyle(
					draggableSnapshot.isDragging,
					draggableProvided.draggableProps.style,
				)}
				>
					{item}
				</div>
			)}
			</Draggable>
		))}
		{droppableProvided.placeholder}
		</div>
	)}
	</Droppable>
</DragDropContext>
```

![Google Chrome - Edit Page ‹ Sandbox — WordPress](https://user-images.githubusercontent.com/24987594/120249574-a9c5a380-c27b-11eb-8c5c-a5e5f61e9aae.gif)

Because of its large feature set and support for horizontal lists, it should be straightforward to implement it as part of the multi-select component as well if it is required.

The documentation is pretty thorough and there are a great number of examples available [here](https://react-beautiful-dnd.netlify.app/?path=/story/single-vertical-list--basic) for almost all use cases.

Of course, this is just my suggestion so please feel free to suggest other alternatives as well or maybe reinforce [dnd-kit](https://github.com/clauderic/dnd-kit) as a stronger option if you would like.

Thank you for taking the time and I look forward to your feedback!

@kuserich @mahdiyazdani @gooklani @kalibir

**Comment by kuserich** [comment]
Thank you very much for writing this discussion and for outlining our options, @AdrianFiroiu . 

`dnd-kit` is smaller (by a lot) although I haven't made a detailed analysis on this (if we include `@dnd-kit/sortable`, for instance, the size advantage already vanishes entirely). However, I believe this is less of an issue in the editor.

Other than that, `react-beautiful-dnd` makes a great contender IMO, particularly because it is backed by Atlassian and it seems that they have put a great amount of thought into this library. Do we expect any use cases other than lists? `react-beautiful-dnd` was specifically built for lists and if we anticipate additional use cases, we should consider a different library.

I also agree with using `react-select` over `react-multi-select-component`.

**Reply by AdrianFiroiu** [reply]
Thank you for the feedback @kuserich and I would especially like to emphasize your point regarding:

> Do we expect any use cases other than lists?

as this should indeed heavily influence our final decision. 

`dnd-kit` is definitely more versatile in this regard - although, for the most part, the range of use cases seem to also revolve around lists and list-type contexts (vertical lists, horizontal lists, grids, multiple containers, nested contexts, variable sized list and grids, transformed items, virtualized lists).

**Comment by mahdiyazdani** [comment]
It appears that the [react-beautiful-dnd](https://github.com/atlassian/react-beautiful-dnd) package doesn’t support _(wrapped)_ grid layouts as of 2021-02-05 as it seems there has been a great and I would say intense discussion around this feature request, the maintainers made it clear that there will no plan in implementing this feature (back) in the future. 

This could be a blocker for us as grid sorting will certainly be needed since we are not intended to install and introduce a sizable library just to be used with lists only.

https://github.com/atlassian/react-beautiful-dnd/issues/316
https://github.com/atlassian/react-beautiful-dnd/issues/1009

**Reply by kuserich** [reply]
Thank you, @mahdiyazdani . That's good to know. I believe this is pretty much a direct requirement from MultiSelect as the list of selected tags is, in essence, a wrapped list.

**Reply by mahdiyazdani** [reply]
I’m in the process of R&D regarding a third-party component that we can rely on and offer as part of our own package to be used with both list and grid items. I will be opening a PR to propose my offering soon.
